### ISSUE TITLE
ECIES definition: ECDH -> KDF issue?

### ISSUE THREAD

**Issue by the-infinity** [ecies/py#151]
Thanks for your great python package! After fiddling a bit around with your library together with others I came to an interoperability question which is maybe a bug, maybe a freedom of implementation (but then I think its worth to document this).

My issue is about encapsulate() and decapsulate(). In line one of encapsulate() you multiply the public key vector with the private key scalar which is found in almost every technical description of ECIES:
> shared_point = peer_public_key.multiply(private_key.secret)

The next line is the one which confuses me: There you concat the shared point with the public key:
```
    master = private_key.public_key.format(compressed=False) + shared_point.format(
        compressed=False
    )
```

I found no other library which does this. After getting deeper into ECIES documentation, I found [this paper](http://www.secg.org/sec1-v2.pdf) where at bottom of page 27 I found this:
```` 
Actions: Calculate a shared secret value as follows:
1. Compute the elliptic curve point P = (xP , yP ) = dUQV .
2. Check that P 6= O. If P = O, output “invalid” and stop.
3. Output z = xP as the shared secret field element.
```` 
Please have a look at the original because copy paste LaTeX rendered docs is really bad. My point is: they use just x_p from shared_point as result, and use this directly in KDF (see page 52 bottom + 53 top).

At [another paper (the ISO propsal)](https://www.shoup.net/papers/iso-2_1.pdf), I found the same strategy at page 26 (15.2 point three). 

So, I'm not an expert on cryptography, so maybe your way is just a degree of freedom you can choose. But to interoperate with libraries which use the way described in the two pdfs it would be great to clearify this. Was your way by intention, and if yes, why? Or is something which should be fixed?

Thanks,
Ernesto

**Comment by kigawas** [comment]
In the ISO proposal you listed:

![image](https://user-images.githubusercontent.com/4182346/87762841-b81af480-c84e-11ea-9613-39df98d6b10f.png)

The `C0 || E'(h~)` is actually concatenating `C0` and `E'(h~)`, which is equal to concat `E(g~, format)` and `E'(h~)`.

You can take E and E' as functions converting a point to hex string or bytes.

![image](https://user-images.githubusercontent.com/4182346/87764186-f1546400-c850-11ea-953d-ce96d168888f.png)

The g is generator, or base point, so `g~=rg` is actually representing the peer's (someone you want to send secret messages to) public key (in our codes it's `shared_point`, exchanged with ephemeral key and peer key) and for the `h~`, which is a random generated key (in our codes it just represents `ephemeral_key` or `ephemeral_public_key` and they are in the first argument in `encap/decap` functions).

For the code above:
```python
master = private_key.public_key.format(compressed=False) + shared_point.format(compressed=False)
                     this is h~                                   this is g~
                  only ephemeral key info                     ephemeral key info and peer public key info
```

Our order is reversed, but at least you have to concat them together and use it in the KDF to generate a shared AES key. And about the format, whether to use compressed or uncompressed format is just a choice.

Hope this answers your question.

**Comment by kigawas** [comment]
You may also check #82 for KDF details.

**Comment by the-infinity** [comment]
Thank you! Yes, I really misunderstood the E0 part.

Because you mentioned the reversed order: just had a look at BountyCastle, they use the same order as you: https://github.com/bcgit/bc-java/blob/master/core/src/main/java/org/bouncycastle/crypto/engines/IESEngine.java#L429 . This seems to be standard somehow. Interesting, I mean ... you have to use the same method on both sides, so different directions are a little strange for me.

But I think the issue can be closed by now. And: I suppose your answer is really good if somebody wants to understand the process behind ECIES, not just use it.

**Comment by the-infinity** [comment]
Ok, there's still one thing, and I suppose that's the core of my confusion. You are absolutely right about the public key which is also part of the KDF input. But I suppose the key issue is that there's no fixed definition if you should use public key plus just the X coordinate or public key plus both coordinates as input for KDF. The first paper describes the way with just one coordinate, the ISO paper the way with both coordinates (I suppose, it's not absolutely clear for me).

If I get it right, in Java BountyCastle they use just one BigInteger = the X part:
> https://github.com/bcgit/bc-java/blob/master/core/src/main/java/org/bouncycastle/crypto/engines/IESEngine.java#L424

In CryptoC++, they also just return one Integer:
> https://cryptopp.com/wiki/Elliptic_Curve_Diffie-Hellman

I suppose in this question there is no "right" or "wrong" way?

**Comment by kigawas** [comment]
The E/E' function in ISO proposals does not specify how you format as long as it's a prefix-free format from my perspective.

Another thing I want to emphasize is that, you should **never** use a simple shared point's x coordinate or (x, y) as your key, even a sha256 hash of it is not safe enough. A correct way is to use HKDF, and the choice is just whether to hash a compressed or uncompressed format in the HKDF.

**Comment by the-infinity** [comment]
Thanks a lot! After some more searching for details I got deeper and deeper into defintions, and stumbled across RFC 6090. [There they write](https://tools.ietf.org/html/rfc6090#section-4.2):

> As described in the final paragraph of [M1985], the x-coordinate of the shared secret value g^(j*k) is a suitable representative for the entire point whenever exponentiation is used as a one-way function.

So I think the core of my issue is "does a library use compact representation?". And maybe the remaining question is: why do you write that it's not safe to use the compact representation? For me, the quote does not include any security concerns. Or maybe there's some current research (RFC 6090 is from 2011) which prooved the RFC-statement wrong?

**Comment by kigawas** [comment]
I'm not saying compressed format is not safe, you can read my comments one more time.

**Comment by the-infinity** [comment]
Oh, sorry. Yes, reading correctly helps to understand. Then I think everything is clear - and I think it's worth mentioning this in documentation, would help to understand interoperability issues.

**Comment by kigawas** [comment]
@the-infinity 
One last  takeaway
> TL;DR Victor Shoup proved that if public key (not only X part, but also Y) is not included into KDF parameters then ECIES is affected to adaptive chosen ciphertext attacks.

https://github.com/ecies/py/issues/82#issuecomment-520192011
